File: 2022/Day6/javascript/Day6.js
// Read File ../input.txt

// Part 1 and Part 2


// ZhengLinLei

const fs = require('fs');

// Read File Async
fs.readFile('../input.txt', 'utf8', (err, text) =>  {
    let fnc = (x, n) => {
        let p = [];
        for (let i = 0; i < x.length; i++) {
            let j = p.indexOf(x[i]);  
            if(j != -1) { 
                p = p.slice(j+1); p.push(x[i]);
            }else{
                if(p.length >= n-1) return i+1
                else p.push(x[i]);
            }       
        }
    }

    console.log("Part 1:", fnc(text.split(''), 4));
    console.log("Part 2:", fnc(text.split(''), 14));
});

File: 2022/Day6/python/Day6.py
# Part 1 and 2
# Zheng Lin Lei


def f(s, n):
    p = ""
    for i, c in enumerate(s):
        j = p.rfind(c)
        if j < 0:
            if len(p) >= n-1:
                return i+1
            else:
                p += c
        else:
            p = p[j+1:]+c
    return 0


t = open("../input.txt", "rt").read().strip()
print(f(t, 14))
